lib/grp.gi

```diff
@@ -4291,86 +4299,125 @@ end );
-                      and HasGeneratorsOfMagmaWithInverses);
+function( gens )
+local G,fam,typ,id,typename;
+
+  fam:=FamilyObj(gens);
+
+  typ:=IsGroup and IsAttributeStoringRep 
+    and HasIsEmpty and HasGeneratorsOfMagmaWithInverses;
+  typename:="defaultGroupType";
+
+  if IsFinite(gens) then
+    typ:=typ and IsFinitelyGeneratedGroup;
+    typename:="defaultFinitelyGeneratedGroupType";
+
+    if Length(gens)>0 and CanEasilyCompareElements(gens) then
+      id:=One(gens[1]);
+      if ForAny(gens,x->x<>id) then
+        typ:=typ and HasIsTrivial and IsNonTrivial;
+        if Length(gens)<=1 then
+          typ:=typ and IsCyclic;
+          typename:="defaultNontrivialCyclicGroupType";
+        else
+          typename:="defaultFinitelyGeneratedNontrivialGroupType";
+        fi;
+      else
+        typ:=typ and IsTrivial and HasIsNonTrivial;
+        typename:="defaultTrivialGroupType";
       fi;
-      typ:=fam!.defaultGroupType;
+    elif Length(gens)<=1 then
+      typ:=typ and IsCyclic;
+      typename:="defaultCyclicGroupType";
     fi;
+  fi;
+  if not IsBound(fam!.(typename)) then
+    typ:=NewType(fam,typ);
+    fam!.(typename):=typ;
+  else
+    typ:=fam!.(typename);
+  fi;
 
-    G:=rec();
-    ObjectifyWithAttributes(G,typ,GeneratorsOfMagmaWithInverses,AsList(gens));
+  G:=rec();
+  ObjectifyWithAttributes(G,typ,GeneratorsOfMagmaWithInverses,AsList(gens));
 
-    return G;
-    end );
+  return G;
+end );
 
 InstallMethod( GroupWithGenerators,
     "generic method for collection and identity element",
     IsCollsElms, [ IsCollection, IsMultiplicativeElementWithInverse ],
-    function( gens, id )
-    local G,fam,typ;
-
-    fam:=FamilyObj(gens);
-    if IsFinite(gens) then
-      if not IsBound(fam!.defaultFinitelyGeneratedGroupWithOneType) then
-        fam!.defaultFinitelyGeneratedGroupWithOneType:=
-          NewType(fam,IsGroup and IsAttributeStoringRep
-                      and HasGeneratorsOfMagmaWithInverses
-                      and IsFinitelyGeneratedGroup and HasOne);
-      fi;
-      typ:=fam!.defaultFinitelyGeneratedGroupWithOneType;
-    else
-      if not IsBound(fam!.defaultGroupWithOneType) then
-        fam!.defaultGroupWithOneType:=
-          NewType(fam,IsGroup and IsAttributeStoringRep
-                      and HasGeneratorsOfMagmaWithInverses and HasOne);
+function( gens, id )
+local G,fam,typ,typename;
+
+  fam:=FamilyObj(gens);
+
+  typ:=IsGroup and IsAttributeStoringRep 
+        and HasIsEmpty and HasGeneratorsOfMagmaWithInverses and HasOne;
+  typename:="defaultGroupWithOneType";
+
+  if IsFinite(gens) then
+    typ:=typ and IsFinitelyGeneratedGroup;
+    typename:="defaultFinitelyGeneratedGroupWithOneType";
+
+    if Length(gens)>0 and CanEasilyCompareElements(gens) then
+      if ForAny(gens,x->x<>id) then
+        typ:=typ and HasIsTrivial and IsNonTrivial;
+        if Length(gens)<=1 then
+          typ:=typ and IsCyclic;
+          typename:="defaultNontrivialCyclicGroupWithOneType";
```

Comment:
BTW, `NEW_TYPE` *also* performs caching of types for a given family. You can inspect `NEW_TYPE_CACHE_MISS` and `NEW_TYPE_CACHE_HIT` before/after creating a group to see it was effective. So.... that might in fact render the cache here moot. (While the explicit cache here may be a tiny bit faster, it also adds a lot of complexity...)

Reply:
I'll put this in a separate remark as a push will make it go outdated in code

Reply:
(after all it did not, but should not be only commit remark)